core/state-machine/src/testing.rs

```diff
@@ -91,81 +117,109 @@ impl<H: Hasher> Default for TestExternalities<H> {
 
 impl<H: Hasher> From<TestExternalities<H>> for HashMap<Vec<u8>, Vec<u8>> {
 	fn from(tex: TestExternalities<H>) -> Self {
-		tex.inner.into()
+		tex.iter_pairs_in_order().collect()
 	}
 }
 
 impl<H: Hasher> From< HashMap<Vec<u8>, Vec<u8>> > for TestExternalities<H> {
 	fn from(hashmap: HashMap<Vec<u8>, Vec<u8>>) -> Self {
-		TestExternalities {
-			inner: hashmap,
-			changes_trie_storage: ChangesTrieInMemoryStorage::new(),
-			changes: Default::default(),
-			code: None,
-		}
+		Self::from_iter(hashmap)
 	}
 }
 
-// TODO child test primitives are currently limited to `changes` (for non child the way
-// things are defined seems utterly odd to (put changes in changes but never make them
-// available for read through inner)
 impl<H: Hasher> Externalities<H> for TestExternalities<H> where H::Out: Ord {
 	fn storage(&self, key: &[u8]) -> Option<Vec<u8>> {
-		match key {
-			CODE => self.code.clone(),
-			_ => self.inner.get(key).cloned(),
-		}
+		self.overlay.storage(key).map(|x| x.map(|x| x.to_vec())).unwrap_or_else(||
+			self.backend.storage(key).expect(EXT_NOT_ALLOWED_TO_FAIL))
 	}
 
 	fn original_storage(&self, key: &[u8]) -> Option<Vec<u8>> {
-		self.storage(key)
+		self.backend.storage(key).expect(EXT_NOT_ALLOWED_TO_FAIL)
```

Comment:
I just copy pasted from core/state-machine/src/ext.rs
But all in all now I think it might event be better to just change test externalities signature and make use of ext directly.. but that would imply adding a few lines in most tests

Reply:
> I just copy pasted from core/state-machine/src/ext.rs
> But all in all now I think it might event be better to just change test externalities signature and make use of ext directly.. but that would imply adding a few lines in most tests

We could add a utility function for the purpose.

Reply:
Yes but it is not as simple as it looks because of how Ext works  https://github.com/paritytech/substrate/issues/2572